File: src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/repository/CaseNoteFilter.java
package uk.gov.justice.hmpps.prison.repository.jpa.repository;

import com.google.common.collect.ImmutableList;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.NoArgsConstructor;
import org.apache.commons.lang3.StringUtils;
import org.jetbrains.annotations.NotNull;
import org.springframework.data.jpa.domain.Specification;
import uk.gov.justice.hmpps.prison.repository.jpa.model.OffenderCaseNote;
import uk.gov.justice.hmpps.prison.service.BadRequestException;
import uk.gov.justice.hmpps.prison.service.QueryParamHelper;

import jakarta.persistence.criteria.CriteriaBuilder;
import jakarta.persistence.criteria.CriteriaQuery;
import jakarta.persistence.criteria.Predicate;
import jakarta.persistence.criteria.Root;
import java.time.LocalDate;
import java.util.List;

@Getter
@Builder()
@AllArgsConstructor
@NoArgsConstructor
@EqualsAndHashCode
public class CaseNoteFilter implements Specification<OffenderCaseNote> {

    private Long bookingId;
    private String prisonId;
    private LocalDate startDate;
    private LocalDate endDate;
    private String type;
    private String subType;
    private List<String> typesSubTypes;

    public static CaseNoteFilterBuilder builder() {
        return new CustomCaseNoteFilterBuilder();
    }

    static class CustomCaseNoteFilterBuilder extends CaseNoteFilterBuilder {
        @Override
       public CaseNoteFilter build() {
            if(super.type != null && super.typesSubTypes != null){
                throw new BadRequestException("Both type and typesAndSubTypes are set, please only use one to filter.");
            }

            return super.build();
        }
    }

    @Override
    public Predicate toPredicate(@NotNull final Root<OffenderCaseNote> root, @NotNull final CriteriaQuery<?> query, @NotNull final CriteriaBuilder cb) {
        final ImmutableList.Builder<Predicate> predicateBuilder = ImmutableList.builder();

        if (bookingId != null) {
            predicateBuilder.add(cb.equal(root.get("offenderBooking").get("bookingId"), bookingId));
        }

        if (StringUtils.isNotBlank(type)) {
            predicateBuilder.add(cb.equal(root.get("type").get("code"), type));
        }

        if (StringUtils.isNotBlank(subType)) {
            predicateBuilder.add(cb.equal(root.get("subType").get("code"), subType));
        }

        if (StringUtils.isNotBlank(prisonId)) {
            predicateBuilder.add(cb.equal(root.get("agencyLocation").get("id"), prisonId));
        }

        if (startDate != null) {
            predicateBuilder.add(cb.greaterThanOrEqualTo(root.get("occurrenceDate"), startDate));
        }

        if (endDate != null) {
            predicateBuilder.add(cb.lessThan(root.get("occurrenceDate"), endDate.plusDays(1)));
        }

        if(typesSubTypes != null && !typesSubTypes.isEmpty()){
           predicateBuilder.add(getTypesPredicate(root, cb));
        }

        final var predicates = predicateBuilder.build();
        return cb.and(predicates.toArray(new Predicate[0]));
    }

    private Predicate getTypesPredicate(final Root<OffenderCaseNote> root, final CriteriaBuilder cb) {
       final var typesAndSubTypes = QueryParamHelper.splitTypes(typesSubTypes);
       final var typesPredicates= typesAndSubTypes.entrySet()
            .stream()
            .map(typeSet -> {
                if(typeSet.getValue().isEmpty()){
                    return cb.equal(root.get("type").get("code"), typeSet.getKey());
                }
                else{
                    return getSubtypesPredicate(root, cb, typeSet.getKey() ,typeSet.getValue());
                }
            })
            .toList();
        return cb.or(typesPredicates.toArray(new Predicate[0]));
    }

    private Predicate getSubtypesPredicate(final Root<OffenderCaseNote> root, final CriteriaBuilder cb, final String type, final List<String> subTypes) {
        final ImmutableList.Builder<Predicate>  typePredicateorBuilder = ImmutableList.builder();
        typePredicateorBuilder.add(cb.equal(root.get("type").get("code"), type));

        final var inTypes = cb.in(root.get("subType").get("code"));
        subTypes.forEach(inTypes::value);
        typePredicateorBuilder.add(inTypes);

        final var typePredicates = typePredicateorBuilder.build();
        return cb.and(typePredicates.toArray(new Predicate[0]));
    }
}

File: src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/repository/CaseNoteFilter.kt
package uk.gov.justice.hmpps.prison.repository.jpa.repository

import jakarta.persistence.criteria.CriteriaBuilder
import jakarta.persistence.criteria.CriteriaQuery
import jakarta.persistence.criteria.Predicate
import jakarta.persistence.criteria.Root
import org.springframework.data.jpa.domain.Specification
import uk.gov.justice.hmpps.prison.repository.jpa.model.OffenderCaseNote
import uk.gov.justice.hmpps.prison.service.BadRequestException
import uk.gov.justice.hmpps.prison.service.QueryParamHelper
import java.time.LocalDate

class CaseNoteFilter(
  val bookingId: Long,
  private val prisonId: String? = null,
  private val startDate: LocalDate? = null,
  private val endDate: LocalDate? = null,
  type: String? = null,
  subType: String? = null,
  typesSubTypes: List<String>? = null,
) : Specification<OffenderCaseNote> {

  private val typesAndSubTypes = if (!type.isNullOrEmpty()) {
    if (!typesSubTypes.isNullOrEmpty()) {
      throw BadRequestException("Both type and typesSubTypes are set, please only use one to filter.")
    }
    if (subType.isNullOrEmpty()) listOf(type) else listOf("$type+$subType")
  } else {
    typesSubTypes
  }

  override fun toPredicate(root: Root<OffenderCaseNote>, query: CriteriaQuery<*>, cb: CriteriaBuilder): Predicate {
    val predicateBuilder = mutableListOf(cb.equal(root.get<Any>("offenderBooking").get<Any>("bookingId"), bookingId))
    if (!prisonId.isNullOrBlank()) {
      predicateBuilder.add(cb.equal(root.get<Any>("agencyLocation").get<Any>("id"), prisonId))
    }
    startDate?.let {
      predicateBuilder.add(cb.greaterThanOrEqualTo(root.get("occurrenceDate"), startDate))
    }
    endDate?.let {
      predicateBuilder.add(cb.lessThan(root.get("occurrenceDate"), endDate.plusDays(1)))
    }
    typesAndSubTypes?.let {
      predicateBuilder.add(getTypesPredicate(root, cb))
    }
    return cb.and(*predicateBuilder.toTypedArray())
  }

  private fun getTypesPredicate(root: Root<OffenderCaseNote>, cb: CriteriaBuilder): Predicate {
    val typesAndSubTypes = QueryParamHelper.splitTypes(typesAndSubTypes)
    val typesPredicates = typesAndSubTypes.entries
      .map { (key, value) ->
        if (value.isEmpty()) {
          cb.equal(root.get<Any>("type").get<Any>("code"), key)
        } else {
          getSubtypesPredicate(root, cb, key, value)
        }
      }

    // if we only have one entry then just return that, which prevents an or clause with only one entry
    if (typesPredicates.size == 1) return typesPredicates[0]

    return cb.or(*typesPredicates.toTypedArray())
  }

  private fun getSubtypesPredicate(
    root: Root<OffenderCaseNote>,
    cb: CriteriaBuilder,
    type: String,
    subTypes: List<String>,
  ): Predicate {
    val typePredicateOrBuilder = mutableListOf(cb.equal(root.get<Any>("type").get<Any>("code"), type))
    val inTypes = cb.`in`(root.get<Any>("subType").get<Any>("code")).apply {
      subTypes.forEach { value(it) }
    }
    typePredicateOrBuilder.add(inTypes)
    return cb.and(*typePredicateOrBuilder.toTypedArray())
  }
}
